comm/p2p/libp2p.go

```diff
@@ -105,42 +114,70 @@ func (c Libp2pCommunication) UnSubscribe(
 
 /** Helper methods **/
 
+func (c Libp2pCommunication) StreamHandlerFunc(s network.Stream) {
+	c.ProcessMessagesFromStream(s)
+	err := s.Close()
+	if err != nil {
+		log.Err(err).Msg("Error closing incoming stream")
+	}
+}
+
+func (c Libp2pCommunication) ProcessMessagesFromStream(s network.Stream) {
+	remotePeerID := s.Conn().RemotePeer()
+	r := bufio.NewReader(s)
+	for {
+		msgBytes, err := ReadStream(r)
+		if err != nil {
+			return
+		}
+
+		var wrappedMsg comm.WrappedMessage
+		if err := json.Unmarshal(msgBytes, &wrappedMsg); nil != err {
+			log.Err(err).Msg("Error unmarshaling message")
+			return
+		}
+		wrappedMsg.From = remotePeerID
+
+		c.logger.Trace().Str(
+			"From", wrappedMsg.From.Pretty()).Str(
+			"MsgType", wrappedMsg.MessageType.String()).Str(
+			"SessionID", wrappedMsg.SessionID).Msg(
+			"processed message",
+		)
+
+		subscribers := c.GetSubscribers(wrappedMsg.SessionID, wrappedMsg.MessageType)
+		for _, sub := range subscribers {
+			sub := sub
+			go func() {
+				sub <- &wrappedMsg
+			}()
+		}
+	}
+}
+
 func (c Libp2pCommunication) sendMessage(
 	to peer.ID,
 	msg []byte,
 	msgType comm.MessageType,
 	sessionID string,
 ) error {
-	pi := c.h.Peerstore().PeerInfo(to)
-	resolver, err := madns.NewResolver()
+	err := c.resolveDNS(to)
 	if err != nil {
 		return err
 	}
-	if len(pi.Addrs) == 0 {
-		return fmt.Errorf("peer %s has no defined addresses", to)
-	}
 
-	addr, err := resolver.Resolve(context.Background(), pi.Addrs[0])
+	var stream network.Stream
+	stream, err = c.streamManager.Stream(sessionID, to)
 	if err != nil {
```

Comment:
`streamManager` will return an error only if it doesn't hold a reference to a stream, but I think that it is possible for the stream to be closed even though there is a reference to the stream object. So I have a feeling that if the stream is closed for some reason (not by invoking `ReleaseStreams()`), this would never recover.

Reply:
Yep, it will error on writing to stream. The problem is you don't really know if a stream is valid or not.
The only thing I can implement is try to create the stream again if write fails. I don't know if that is overkill as the stream is per session and if stream fails (relayer is offline) that session can't be repeated anyway, but probably it could help in some cases.

Reply:
Also not sure how helpful this would be as it is hard to evaluate this edge-case without seeing how it behaves on deployment. Your call; we can always monitor if this case is happening and refactor accordingly.

Reply:
I would leave this for now and then see how it goes. The only way I manage to reproduce is to shut off a relayer, but then I can't really do much to recover anyway.